### PR TITLE
Introduce globals

### PR DESCRIPTION
**Pull request by devoncarew**
- change the global `AppContext context` variable from knowing about specific services to being a type -> object map for a zone. Each zone can introduce their own local overrides. This is useful for testing; each `context.runInZone()` call essentially becomes a container with it's own values for the globals, so you can use mocks for testing w/o changing the call locations.
- introduce some globals for the Logger and DeviceManager class (in the `base/globals.dart` library).
- create a `testUsingContext()` method to define a test, and run that test in a new context with some default mocks set up

Hopefully using these zone-specific globals will allow us to have fewer statics, pass fewer objects along method call chains, and make testing easier.

### CODE REVIEW

File: packages/flutter_tools/lib/src/base/context.dart [hunk 1]
```diff
@@ -14,57 +13,43 @@ AppContext get context {
...
-
-  void printError(String message, [StackTrace stackTrace]);
-  void printStatus(String message);
-  void printTrace(String message);
-}
-
-class _DefaultAppContext implements AppContext {
-  DateTime _startTime = new DateTime.now();
+    AppContext parent = _calcParent(Zone.current);
+    return parent != null ? parent.getVariable(type) : null;
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/base/context.dart]: you can use ?. here

**Comment by Hixie** [comment]
squash your commits before checkin

**Comment by Hixie** [comment]
lgtm at the line level, but i don't really understand the overall architecture you're going for here.

File: packages/flutter_tools/lib/src/base/context.dart (hunk 1 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/src/base/context.dart]: done

**Comment by devoncarew** [comment]
I hope to simplify some code with this going forward, and open the door to some refactorings, like perhaps  separating some of the app functionality out of `FlutterCommand`.

**Comment by Hixie** [comment]
I look forward to seeing the results ☺

**Pull request merged**